lib/validator/validator.go

```diff
@@ -286,19 +285,32 @@ func validateHostEndpointSpec(v *validator.Validate, structLevel *validator.Stru
 	}
 }
 
-func validatePoolMetadata(v *validator.Validate, structLevel *validator.StructLevel) {
-	pool := structLevel.CurrentStruct.Interface().(api.IPPoolMetadata)
-
-	// The Calico IPAM places restrictions on the minimum IP pool size, check that the
-	// pool is at least the minimum size.
-	if pool.CIDR.IP != nil {
-		ones, bits := pool.CIDR.Mask.Size()
-		log.Debugf("Pool CIDR: %s, num bits: %d", pool.CIDR, bits-ones)
-		if bits-ones < 6 {
-			structLevel.ReportError(reflect.ValueOf(pool.CIDR),
-				"CIDR", "", reason("IP pool is too small"))
+func validateIPPool(v *validator.Validate, structLevel *validator.StructLevel) {
+	pool := structLevel.CurrentStruct.Interface().(api.IPPool)
+
+	// Validation of the data occurs before checking whether Metadata
+	// fields are complete, so need to check whether CIDR is assigned before
+	// performing cross-checks.  If CIDR is not assigned this will be
+	// picked up during Metadata->Key conversion.
+	if pool.Metadata.CIDR.IP != nil {
+		// IPIP cannot be enabled for IPv6.
+		if pool.Metadata.CIDR.Version() == 6 && pool.Spec.IPIP != nil && pool.Spec.IPIP.Enabled {
+			structLevel.ReportError(reflect.ValueOf(pool.Spec.IPIP.Enabled),
+				"IPIP.Enabled", "", reason("IPIP is not supported on an IPv6 IP pool"))
+		}
+
+		// The Calico IPAM places restrictions on the minimum IP pool size.  If
+		// the pool is enabled, check that the pool is at least the minimum size.
+		if !pool.Spec.Disabled {
+			ones, bits := pool.Metadata.CIDR.Mask.Size()
+			log.Debugf("Pool CIDR: %s, num bits: %d", pool.Metadata.CIDR, bits-ones)
+			if bits-ones < 6 {
+				structLevel.ReportError(reflect.ValueOf(pool.Metadata.CIDR),
+					"CIDR", "", reason("IP pool is too small"))
```

Comment:
Might be useful if the error message indicates that it's too small for an IPAM enabled Pool.